### PR TITLE
Replace some options with nullable vars

### PR DESCRIPTION
**Pull request by zyedidia**
This change improves memory usage by replacing Option objects with nullable vars that can be optimized by the JIT. Benchmarks show that memory usage is reduced by roughly 6%.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- performance improvement

#### API Impact

Does not change the current API.

#### Backend Code Generation Impact

Does not change the code generation.

#### Desired Merge Strategy

Squash.

#### Release Notes

Internal performance improvement that reduces memory usage by roughly 6%.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Review by mwachs5 — COMMENTED** [review]
since we generally discourage the use of `var`s and `null`, should we put a brief comment on these usages e.g. "using nullable vars for better memory usage"?

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 1]
```diff
@@ -87,10 +87,28 @@ trait InstanceId {
 }
 
 private[chisel3] trait HasId extends InstanceId {
-  private[chisel3] var _parent: Option[BaseModule] = Builder.currentModule
+  // using nullable var for better memory usage
+  private var _parentVar: BaseModule = Builder.currentModule match {
+    case Some(m) => m
+    case None    => null
+  }
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:94]:
```suggestion
  private var _parentVar: BaseModule = Builder.currentModule.getOrElse(null)
```

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 2]
```diff
@@ -87,10 +87,28 @@ trait InstanceId {
...
+  private[chisel3] def _parent: Option[BaseModule] = Option(_parentVar)
+  private[chisel3] def _parent_=(target: Option[BaseModule]) {
+    _parentVar = target match {
+      case Some(m) => m
+      case None    => null
+    }
+  }
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:101]:
```suggestion
  private[chisel3] def _parent_=(target: Option[BaseModule]): Unit = {
    _parentVar = target.getOrElse(null)
  }
```
Pithier way to express it, also avoid procedure syntax (always have an `=` when defining a method).

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 3]
```diff
@@ -87,10 +87,28 @@ trait InstanceId {
...
   // Set if the returned top-level module of a nested call to the Chisel Builder, see Definition.apply
-  private[chisel3] var _circuit: Option[BaseModule] = None
+  private var _circuitVar:       BaseModule = null // using nullable var for better memory usage
+  private[chisel3] def _circuit: Option[BaseModule] = Option(_circuitVar)
+  private[chisel3] def _circuit_=(target: Option[BaseModule]) {
+    _circuitVar = target match {
+      case Some(m) => m
+      case None    => null
+    }
+  }
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:111]:
```suggestion
  private[chisel3] def _circuit_=(target: Option[BaseModule]) {
    _circuitVar = target.getOrElse(null)
  }
```

**Review by jackkoenig — COMMENTED** [review]
Overall looks great, thanks for doing this and for measuring the improvement?

I made a few nitty suggetsions.

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 4]
```diff
@@ -200,11 +211,12 @@ private[chisel3] trait HasId extends InstanceId {
       naming_prefix = Nil
     }
 
-  private var _ref: Option[Arg] = None
+  private var _refVar: Arg = null
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:214]:
```suggestion
  private var _refVar: Arg = null // using nullable var for better memory usage
```

**Review by jackkoenig — APPROVED** [review]

**Pull request merged**